rust/src/daemon.rs

```diff
@@ -131,6 +144,146 @@ fn deployment_populate_variant_origin(
     Ok(())
 }
 
+// Serialize to JSON and write to the stream, then close it.  There's no framing, so
+// closing the socket signals end of message.
+#[context("Sending activation message")]
+pub(crate) fn write_message(conn: &mut UnixStream, message: SocketMessage) -> Result<()> {
+    let sendbuf = serde_json::to_vec(&message)?;
+    conn.write_all(&sendbuf)?;
+    // Flush writes
+    conn.shutdown(std::net::Shutdown::Write)?;
+    Ok(())
+}
+
+// Receive and parse a single message from the socket.
+#[context("Receving activation message")]
+pub(crate) fn recv_message(conn: &mut UnixStream) -> Result<SocketMessage> {
+    let mut buf = Vec::new();
+    conn.read_to_end(&mut buf)?;
+    conn.shutdown(std::net::Shutdown::Read)?;
+    let msg: SocketMessage = serde_json::from_slice(&buf).context("Parsing client message")?;
+    Ok(msg)
+}
+
+fn client_hello(mut client: UnixStream, e: anyhow::Result<()>) -> Result<()> {
+    let msg = recv_message(&mut client)?;
+    let reply = match msg {
+        SocketMessage::ClientHello { selfid } => {
+            let myid = crate::core::self_id()?;
+            if selfid != myid {
+                // For now, make this not an error
+                tracing::warn!("Client reported id: {selfid} different from mine: {myid}");
+            }
+            match e {
+                Ok(()) => SocketMessage::ServerOk,
+                Err(e) => SocketMessage::ServerError {
+                    msg: format!("{e}"),
+                },
+            }
+        }
+        o => SocketMessage::ServerError {
+            msg: format!("Unexpected message: {o:?}"),
+        },
+    };
+    write_message(&mut client, reply).context("Writing client reply")?;
+    tracing::debug!("Acknowleged client");
+    Ok(())
+}
+
+static SHUTDOWN_SIGNAL: Lazy<Mutex<Option<Sender<()>>>> = Lazy::new(|| Mutex::new(None));
```

Comment:
I was looking at this again and I remembered that there's a race condition this was intended to fix but we didn't.  Basically, the socket clients aren't accounted for in our current exit-on-idle logic.

What we need to do is:

- If we have a socket client, then disable the idle exit
- When we choose to idle exit, first shut down the `accept()` handler

Reply:
Mmm and a further complication here is to make this race-free we probably need to use the glib main loop (aka main thread) and not the tokio thread - at least to *accept* clients on the socket; we could still do the actual processing on the tokio thread with appropriate synchronization.